src/Z3/RLock.hs

```diff
@@ -0,0 +1,122 @@
+    deriving Eq
+{-| The state of an 'RLock'.
+
+* 'Nothing' indicates an \"unlocked\" state.
+
+* @'Just' (tid, n)@ indicates a \"locked\" state where the thread identified by
+@tid@ acquired the lock @n@ times.
+-}
+type State = Maybe (ThreadId, Integer)
+
+-- | Create a reentrant lock in the \"unlocked\" state.
+new :: IO RLock
+new = do lock <- Lock.new
+         RLock <$> newMVar (Nothing, lock)
+
+{-|
+Acquires the 'RLock'. Blocks if another thread has acquired the 'RLock'.
+@acquire@ behaves as follows:
+
+* When the state is \"unlocked\", @acquire@ changes the state to \"locked\"
+with the current thread as owner and an acquired count of 1.
+
+* When the state is \"locked\" and the current thread owns the lock @acquire@
+only increments the acquired count.
+
+* When the state is \"locked\" and the current thread does not own the lock
+@acquire@ /blocks/ until the owner releases the lock. If the thread that called
+@acquire@ is woken upon release of the lock it will take ownership and change
+the state to \"locked\" with an acquired count of 1.
+
+There are two further important properties of @acquire@:
+
+* @acquire@ is single-wakeup. That is, if there are multiple threads blocked on
+@acquire@, and the lock is released, only one thread will be woken up. The
+runtime guarantees that the woken thread completes its @acquire@ operation.
+
+* When multiple threads are blocked on @acquire@ they are woken up in FIFO
+order. This is useful for providing fairness properties of abstractions built
+using locks. (Note that this differs from the Python implementation where the
+wake-up order is undefined.)
+-}
+acquire :: RLock -> IO ()
+acquire (RLock mv) = do
+  myTID <- myThreadId
+  mask_ $ let acq = do t@(mb, lock) <- takeMVar mv
+                       case mb of
+                         Nothing          -> do Lock.acquire lock
+                                                putMVar mv (Just (myTID, 1), lock)
+                         Just (tid, n)
+                           | myTID == tid -> let !sn = succ n
+                                             in putMVar mv (Just (tid, sn), lock)
+                           | otherwise    -> do putMVar mv t
+                                                Lock.wait lock
+                                                acq
+          in acq
+
+{-| @release@ decrements the acquired count. When a lock is released with an
+acquired count of 1 its state is changed to \"unlocked\".
+
+Note that it is both an error to release a lock in the \"unlocked\" state and to
+release a lock that is not owned by the current thread.
+
+If there are any threads blocked on 'acquire' the thread that first called
+@acquire@ will be woken up.
+-}
+release :: RLock -> IO ()
+release (RLock mv) = do
+  myTID <- myThreadId
+  mask_ $ do
+    t@(mb, lock) <- takeMVar mv
+    let err msg = do putMVar mv t
+                     error $ "Control.Concurrent.RLock.release: " ++ msg
+    case mb of
```

Comment:
Perhaps fix the error message, `s/Control.Concurrent.RLock.release/Z3.RLock.release`.